GVFS/GVFS.Service/Handlers/RequestHandler.cs

```diff
@@ -0,0 +1,87 @@
+﻿using GVFS.Common.NamedPipes;
+using GVFS.Common.Tracing;
+using System.Runtime.Serialization;
+
+namespace GVFS.Service.Handlers
+{
+    public class RequestHandler
+    {
+        private string etwArea;
+        private ITracer tracer;
+        private RepoRegistry repoRegistry;
+
+        public RequestHandler(ITracer tracer, string etwArea, RepoRegistry repoRegistry)
+        {
+            this.tracer = tracer;
+            this.etwArea = etwArea;
+            this.repoRegistry = repoRegistry;
+        }
+
+        public void HandleRequest(ITracer tracer, string request, NamedPipeServer.Connection connection)
+        {
+            NamedPipeMessages.Message message = NamedPipeMessages.Message.FromString(request);
+            if (string.IsNullOrWhiteSpace(message.Header))
+            {
+                return;
+            }
+
+            using (ITracer activity = this.tracer.StartActivity(message.Header, EventLevel.Informational, new EventMetadata { { "request", request } }))
+            {
+                string deserializeError = null;
+
+                try
+                {
+                    this.HandleMessage(activity, message, connection, out deserializeError);
+                }
+                catch (SerializationException ex)
+                {
+                    activity.RelatedError("{0}: {1}", deserializeError, ex.Message);
+                }
+            }
+        }
+
+        protected virtual void HandleMessage(
+            ITracer tracer,
+            NamedPipeMessages.Message message,
+            NamedPipeServer.Connection connection,
+            out string deserializeError)
+        {

```

Comment:
Sorry one more thought I have on this is using the `MessageHandler` base class a little more for some of this code adding some code.
```
    public abstract class MessageHandler
    {
        public abstract string DeserializeErrorMessage { get; } 
        public abstract void Run(NamedPipeMessages.Message message);

        protected void WriteToClient(NamedPipeMessages.Message message, NamedPipeServer.Connection connection, ITracer tracer)
        {
            if (!connection.TrySendResponse(message))
            {
                tracer.RelatedError("Failed to send line to client: {0}", message);
            }
        }
    }
```
This will make each `MessageHandler` object responsible to create its own request object in the `Run` method and have its own specific message for deserialization errors and the request object would not need to be created and passed to the constructor.

Then have a method for creating the specific message handler that would be what gets overridden in the Windows class
```
public virtual MessageHandler CreateMessageHandler(ITracer tracer, RepoRegistry repoRegistry, NamedPipeServer.Connection connection, string messageHeader)
{
    switch (messageHeader)
    {
        case NamedPipeMessages.RegisterRepoRequest.Header:
            return new RegisterRepoHandler(tracer, this.repoRegistry, connection);
        ...
        default:
            return null;
    }
}
```

Then this method would go away and you would have this in the `HandleRequest`
```
    MessageHandler handler = this.CreateMessageHandler(...);
    if (handler == null)
    {
        <log the unknown request>
        return;
    }

    try
    {
        handler.Run(message);
    }
    catch (SerializationException ex)
    {
        activity.RelatedError("{0}: {1}", handler.DeserializeErrorMessage, ex.Message);
    }
```

I hope that makes sense and let me know if you think this is abstracting things too much.  The thing I didn't like was the `out` for the `deserializeError` and how much it needed to be set.

Reply:
Also looks like we can move at least the `ITracer tracer` and `NamedPipeServer.Connection connection` into the base class and constructor since all the derived classes have them.

Reply:
Couple of issues there.
1. Refactoring of existing code - Most of this code has been existing in GVFS.Service, only re-arranged for `Service` porting to Mac. I have created an [issue](https://github.com/Microsoft/VFSForGit/issues/1032) to track the refactoring as it is not directly related to `Service` porting effort.
2. As discussed - removed the usage of deserializeError, added a new request descriptor that is used in error logging.